---
title: "TRIFID_using_extra_introns"
output: pdf_document
---
Makes Supplementary Table 5.
trifid volcano plot - Figure2F

Required for:
Trifid_stats.Rmd - Figure 2E
Go_networks_plus_trifid.Rmd - Figure 2G,H

Theres a refseq TRIFID table, so for introns with only a refseq id I use the TRIFID score for refseq transcripts.  


```{r message=FALSE, warning=FALSE}
library(biomaRt)
library(tidyr)
library(dplyr)
library(ggplot2)
library(ggrepel)

library(clusterProfiler)
library(here)
i_am("R/15_TRIFID_using_extra_introns.Rmd")
```


```{r}
lc = read.delim(here("31_leafcutter", "three_database_info_all_junctions.tsv"))
lc = unite(lc, "intron_coords", chr, start, end, strand, sep = ":")
#lc[grep("PEMT",lc$gene),]
head(lc)
dim(lc) # some duplications based on gene names/ antisense transcripts.  
```

```{r}
trifid = read.delim(here("annotations/trifid/gencode37_trifid_predictions.tsv"))
head(trifid)
length(unique(trifid$transcript_id))#104 688

trefseq = read.delim(here("annotations/trifid/refseq110_trifid_predictions.tsv"))
nrow(trefseq)
head(trefseq)
trefseq$gene_id = as.character(trefseq$gene_id)
trifid = bind_rows(trifid, trefseq)
nrow(trifid)
```

The reason trifid has so few transcripts is only those with translated sequences are included.  

## Convert lc introns to transcript table

```{r}
transcripts = separate_longer_delim(lc, transcript_ids, delim=",")
transcripts$transcript_ids = gsub("\\.[0-9]*","",gsub("rna-","",transcripts$transcript_ids))
head(transcripts) 
dim(transcripts)

#transcripts[grep("PEMT",transcripts$gene),]
```
 some numbers on this
```{r}
sig_trans = filter(transcripts, p.adjust < 0.05 & abs(deltapsi) >= 0.1)
sprintf("Leafcutter significant introns (%d) correspond to %d unique transcripts (represented in %d rows).",
        length(unique(sig_trans$intron_coords)),
        length(unique(sig_trans$transcript_ids)),
        length(sig_trans$transcript_ids))

sprintf("%d of these transcripts are represented in the trifid database.\n",
        sum(unique(sig_trans$transcript_ids) %in% trifid$transcript_id))

transcripts$transcript_ids[grep("PEMT",transcripts$gene,)] %in% trifid$transcript_id
```


## Merge trifid and leafcutter

```{r}
mtrans = merge(transcripts, trifid, by.x ="transcript_ids",by.y="transcript_id")
mtrans = mutate(mtrans, condition = if_else(deltapsi > 0, "beige", if_else(deltapsi < 0, "white", "none") ),
                sig = p.adjust < 0.05 & abs(deltapsi) >= 0.1)

length(unique(mtrans$transcript_ids)) #82 326 transcripts in total
length(unique(filter(mtrans, sig) %>% pull(transcript_ids))) 
```

```{r}
mtrans = arrange(mtrans, desc(trifid_score), desc(norm_trifid_score), p.adjust, desc(abs(deltapsi))) 
head(mtrans[c("gene_name", "trifid_score","norm_trifid_score","deltapsi","p.adjust")], n=20)

#head(mtrans[mtrans$gene == "PEMT",])
```
## Log missing introns

```{r}
missing = transcripts[!transcripts$transcript_ids %in% trifid$transcript_id &
                        !transcripts$intron_coords %in% mtrans$intron_coords,] #annotate later

cat("Unrepresented transcripts include those from genes like: \n")
cat(head(unique(missing$gene[missing$p.adjust < 0.05 & abs(missing$deltapsi)>0.1])))
table(missing$annotation)
missing[grep( "PEMT", missing$gene),] #Missing PEMT transcripts from non-significant cluster
```

Of course  fantom and cryptic transcripts cannot be annotated by this trifid (which is based on the ensembl + refseq annotation); but also many transcripts from gencode cannot be found there either - perhaps they're new transcripts, or not protein coding versions.  The genocde annotation is always changing.  

```{r}
table(mtrans$condition)
table(paste(mtrans$sig, mtrans$condition))
```

## Histogram

```{r}
ggplot(filter(mtrans, sig)) + geom_histogram(aes(x=norm_trifid_score, fill=condition), bins=30)
```
## Flags

```{r}
table(mtrans$flags)
table(filter(mtrans, sig) %>% pull(flags))
```
## Correlation between dPSI and TRIFID score?

There is actually, if you look within the same cluster if we have a intron with a higher PSI it tends to have higher functionality - or maybe it just has more transcripts?  
```{r}
ggplot(mtrans, aes(x=deltapsi, y=norm_trifid_score, colour=condition)) + geom_point(size=0.5) +
    geom_text_repel(data= filter(mtrans, abs(deltapsi) >0.3), aes(label=gene))
ggplot(mtrans, aes(x=abs(deltapsi), y=norm_trifid_score, colour=condition)) + geom_point(size=0.5) +
    geom_text_repel(data= filter(mtrans, abs(deltapsi) >0.3), aes(label=gene))

```

## Annotate transcript names

```{r}
mart <- useMart(biomart = "ensembl",
  dataset = "hsapiens_gene_ensembl",
   host = "https://sep2019.archive.ensembl.org")

biomaRt::searchAttributes(mart = mart, pattern = "transcript.*name")

annot = getBM(c("external_transcript_name","ensembl_gene_id", "ensembl_transcript_id"), 
              filters = "ensembl_transcript_id",
              values =  mtrans$transcript_ids,
              mart = mart, useCache = F)

head(annot, n=2); dim(annot)

#Add gene names to filt_series
mtrans = merge(mtrans,annot, by.x= "transcript_ids", 
                        by.y = "ensembl_transcript_id", sort=FALSE,
               all.x = T)
#head(mtrans)
remove(annot)
```


## Averaging score across isoforms at each intron

```{r}
summary(mtrans$trifid_score)

head(mtrans)
dim(mtrans)
#mtrans = filter(mtrans, sig)

#unknown gene names get given a ".", change to something more meaningful
mtrans = mutate(mtrans, gene = if_else(gene==".", cluster_id, gene))

#this step  averages the score if multiple transcripts are implicated at an intron
introns = group_by(mtrans, intron_coords,condition, deltapsi, p.adjust, gene, cluster_id, annotation) %>% summarise(mean_trifid_score = mean(trifid_score, na.rm = T),
                                                                 mean_norm_score = mean(norm_trifid_score, na.rm = T),
                                                                  median_norm_score = median(norm_trifid_score, na.rm = T),
                                                                 transcripts = paste(transcript_ids, sep=",",collapse=","),
                                                                 transcript_names = paste(external_transcript_name,sep=",", collapse=","),) 
introns = arrange(introns, desc(abs(deltapsi)), desc(mean_trifid_score), p.adjust,)
dim(introns) #introns with annotations
head(introns)
length(unique(introns$intron_coords))
```

## If an intron is not in trifid; report as -0.1
```{r}
sig_clusters = unique(lc$cluster_id[lc$p.adjust < 0.05 & abs(lc$deltapsi) > 0.1 & lc$annotation %in% c("gencode","refseq")])
summary(sig_clusters %in% introns$cluster_id)#420 clusters have at least 1 trifid annotation; 78 do not
#must have another annotated intron in the cluster, but not have a score already for that intron
to_add = missing#[missing$cluster_id %in% sig_clusters & 
                 #    !missing$intron_coords %in% introns$intron_coords,]
nrow(to_add) #unannotated trifid transcripts, have an annotated cluster and are for an unscored intron

to_add = mutate(to_add, gene = if_else(gene==".", cluster_id, gene))
to_add = group_by(to_add, intron_coords, deltapsi, p.adjust, gene, cluster_id, annotation) %>% 
    summarise(transcripts = paste(transcript_ids,collapse=","),
              ) 
to_add = mutate(to_add, mean_trifid_score = -0.1,
                        mean_norm_score = -0.1,
                        median_norm_score = -0.1,
                condition=  if_else(deltapsi > 0, "beige", "white"))
head(to_add)
nrow(to_add)
```

```{r}
all_introns = bind_rows(in_trifid=introns, not_in_trifid=to_add[to_add$cluster_id %in% introns$cluster_id,],
                        cluster_not_in_trifid= to_add[!to_add$cluster_id %in% introns$cluster_id,],
                        .id = "in_trifid" )
nrow(all_introns)
length(unique(all_introns$intron_coords))
table(all_introns$in_trifid)
table(filter(all_introns, p.adjust < 0.05 & abs(deltapsi) > 0.1 & annotation %in% c("gencode","refseq")) %>% pull(in_trifid))
```
Gencode + Refseq = 647 + 33 = 680 junctions; 
of which 577 are in trifid; 29 are  not in trifid; 
and 74 of which none of the junctions in the cluster are in trifid, e.g. 

```{r}
filter(all_introns, in_trifid == "cluster_not_in_trifid"& p.adjust < 0.05 & abs(deltapsi) > 0.1) %>% 
  arrange(gene) %>% pull(gene) %>% unique()

filter(all_introns, p.adjust < 0.05 & abs(deltapsi) > 0.1 & annotation %in% c("gencode","refseq")) %>% 
  pull(cluster_id) %>% unique() %>% length()

filter(all_introns, p.adjust < 0.05 & abs(deltapsi) > 0.1 & annotation %in% c("gencode","refseq") & in_trifid == "cluster_not_in_trifid") %>% 
  pull(cluster_id) %>% unique() %>% length()

filter(all_introns, p.adjust < 0.05 & abs(deltapsi) > 0.1 & annotation %in% c("gencode","refseq") & in_trifid != "cluster_not_in_trifid") %>% 
  pull(cluster_id) %>% unique() %>% length()
```


## Violin plots on *average* TRIFID score

```{r}
all_introns =mutate(all_introns, sig = if_else(p.adjust < 0.05 & deltapsi > 0.1, "sig_beige", 
                                       if_else(p.adjust < 0.05 & deltapsi < -0.1, "sig_white",
                                                       if_else(deltapsi > 0, "beige", 
                                                               if_else(deltapsi < -0, "white", "neither")))) )
table(all_introns$sig)

ggplot(all_introns, aes(x=sig, y=mean_norm_score, fill=sig)) + geom_violin() +
    geom_jitter(size=0.75)
ggplot(all_introns, aes(x= sig,  y=mean_norm_score,colour=sig)) + geom_boxplot() +
    geom_jitter(size=0.75)
```
The non-averaged score would be appropriate if the number of transcripts per intron was similar between the groups.  Trifid scores (more appropriate if we're comparing across gene instead of pairwise between within each gene); has the average score HIGHER in white than beige.  also a nice validation of the psi threshold.  

```{r}
all_introns[grepl("NDUF", all_introns$gene) & all_introns$p.adjust < 0.05 & abs(all_introns$deltapsi)>0.1,]
write.table(all_introns, here("31_leafcutter/trifid_all_introns.tsv"), sep="\t",quote=F, row.names = F)
```


## Select significant introns + alt introns

```{r}
alt_introns = read.delim(here("31_leafcutter/alt_introns_195.tsv"))
alt_introns = unite(alt_introns, "intron_coords", chr, start, end, strand, sep = ":")
sig_junctions = filter(all_introns, (p.adjust < 0.05 & abs(deltapsi) > 0.1 ) |
                                    intron_coords %in% alt_introns$intron_coords) %>% arrange(p.adjust)
nrow(sig_junctions)
table(sig_junctions$sig)
```

```{r}
write.table(sig_junctions, here("31_leafcutter/trifid_with_alt_introns.tsv"), sep="\t",quote=F, row.names = F)
```


## Plot the TRIFID difference

```{r}
ggplot(sig_junctions, aes(colour=condition, y=mean_norm_score, x=deltapsi)) + geom_jitter() + 
    geom_line(data=filter(sig_junctions, abs(deltapsi) > 0.25), aes(group=cluster_id), colour="black", alpha=0.5) + 
    geom_text_repel(data=filter(sig_junctions, abs(deltapsi) > 0.25), aes(label=gene))
```

## Calculate the TRIFID difference!!

```{r}
#if three introns are signifciant for a cluster
#average TRIFID score between the two introns in the same direction 

paste_uq = function(x){
    return = paste(unique(x), collapse=",")
}
nrow(filter(sig_junctions, in_trifid != "cluster_not_in_trifid"))
mean_of_3s = filter(sig_junctions, in_trifid != "cluster_not_in_trifid") %>%
    group_by(cluster_id, condition, p.adjust) %>% summarise(mean_norm_score = mean(mean_norm_score),
                                                                deltapsi = mean(deltapsi),
                                                                gene = paste(unique(gene), collapse=","),
                                                                transcript_names= paste(unique(transcript_names), collapse=","))
head(mean_of_3s)
nrow(mean_of_3s)#15 clusters must be averaged
#filter(mean_of_3s, is.na(deltapsi))

trifid_diff = pivot_wider(mean_of_3s, names_from = condition, values_from = c(deltapsi, mean_norm_score, gene), values_fn = list(deltapsi=mean,mean_norm_score=mean, gene=paste_uq ),
                          id_cols=c("cluster_id", "p.adjust"))
head(trifid_diff); nrow(trifid_diff)#420 clusters - 
#should be gencode + refseq - cluster_not_in_trifid 
#470  - 58 + alt_genocde/refseq 8 = 420

```

```{r}
## If your other intron is not in trifid speak now
## deltapsi we should have from the other table, just the trifid score we could set to -1
filter(trifid_diff, is.na(deltapsi_beige ) | is.na(deltapsi_white )) # just 2 with an intron pair missing
sig_junctions[sig_junctions$cluster_id %in%c("clu_30508_+","clu_3320_-"),] #beacuse the deltapsi goes in the same direction

trifid_diff = filter(trifid_diff,!is.na(deltapsi_beige ) & !is.na(deltapsi_white )) 

```


```{r}
trifid_diff = mutate(trifid_diff, trifid_diff = mean_norm_score_beige-mean_norm_score_white, max_dpsi = max(abs(c(deltapsi_white, deltapsi_beige))),
                     gene_to_plot = if_else(trifid_diff > 0, gene_beige, gene_white))
head(trifid_diff)
```


```{r}
figs = here("R/plots")
ggplot(trifid_diff, aes(y=max_dpsi, x=trifid_diff, colour=trifid_diff)) + geom_point() + 
    geom_label_repel(data=filter(trifid_diff, max_dpsi > 0.25), aes(label=gene_to_plot)) + 
    scale_color_gradient2(low="turquoise", mid="black", high="red")
ggsave(file.path(figs, "trifid_difference_v_dpsi.pdf"))

ggplot(trifid_diff, aes(y=-log10(p.adjust), x=trifid_diff, colour=trifid_diff)) + geom_point() + 
    geom_label_repel(data=filter(trifid_diff, p.adjust < 0.01), aes(label=gene_to_plot)) + 
    scale_color_gradient2(low="turquoise", mid="black", high="red")
ggsave(file.path(figs, "trifid_difference_v_pvalue.pdf"))
```

```{r}
write.table(trifid_diff, here("31_leafcutter/trifid_DIFFERENCE_with_Alt_introns.tsv"), sep="\t",quote=F, row.names = F)
```


## GSEA
### Setup

```{r}
molsig <- clusterProfiler::read.gmt(here("annotations","msigdb.v2023.1.Hs.symbols.gmt"))
head(molsig); nrow(molsig)
prefixes = c("HALLMARK", "KEGG","REACTOME", "WP", "GOBP", "GOCC","GOMF")
colnames(molsig) = c("term","gene")
some.molsig = molsig[gsub("_.*","", molsig$term) %in% prefixes,]
some.molsig$term = factor(some.molsig$term)
table(gsub("_.*","", some.molsig$term))
rm(molsig)

shorten = function(ont) {
  abbreviate(gsub("_"," ", tolower(ont)),minlength=40, dot=T, named = F)
}

```


```{r}
genelist = trifid_diff$trifid_diff
names(genelist) = trifid_diff$gene_to_plot
genelist = genelist[order(genelist, decreasing = T)]
length(genelist)
summary(genelist)
gse = GSEA(genelist, TERM2GENE= some.molsig, pvalueCutoff=1)
head(gse)
```
```{r}
dotplot(gse, showCategory = 10, x="enrichmentScore")
cnetplot(gse, showCategory=3, categorySize="pvalue", color.params = list(edge=T))
to_print = gse[,c("ID","setSize", "enrichmentScore"  ,"NES",
                  "pvalue"    ,  "p.adjust" ,   "qvalue"    ,  "core_enrichment"   ,   "rank"  )]
write.table(to_print, here("31_leafcutter","TRIFID_GSEA.txt"))
```


```{r}
custom_ora_to_df = function(res, annot=NULL, other_cols=NULL){
    res_df = res[,c(other_cols,"ID","setSize", "enrichmentScore"  ,"NES",
                  "pvalue"    ,  "p.adjust" ,   "qvalue"    ,  "core_enrichment"   ,   "rank" )]

    print(dim(res_df))
    if (length(annot)>1){
        res_df = merge(res_df, annot, by.x="ID", by.y="term", sort=F)
    }
    
    res_df = res_df[order(res_df$p.adjust),]
    return(res_df)
}
```


```{r}
sep_go = list()
for (db in prefixes){
    print(db)
    t2g = some.molsig[grep(db, some.molsig$term),]
    ea = GSEA(genelist, TERM2GENE = t2g, pvalueCutoff = 1,minGSSize = 3)
    df = custom_ora_to_df(ea)
    print(head(df[2:8], n=10))
    sep_go[[db]] = ea
    #print(dotplot(ea, showCategory=20) +ggtitle(db))
    #print(cnetplot(ea, geneSetID = 1:5) + ggtitle(db))
}
```




