---
title: "trifid_stats"
output: html_document
date: "2023-12-15"
---

```{r message=FALSE, warning=FALSE}
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggpubr)
```

```{r}
LEAF= "C:/Users/sarahhp/OneDrive - Universitetet i Oslo/Projects/rnaseq/beige_rnaseq_jul22/31_leafcutter"
figs = "C:/Users/sarahhp/OneDrive - Universitetet i Oslo/Projects/rnaseq/beige_rnaseq_jul22/R/leafcutter/plots"
old_trifid = read.delim(file.path(LEAF, "annotation", "trifid_summary_per_intron.tsv"))
nrow(old_trifid)
table(old_trifid$sig)

new_trifid =  read.delim(file.path(LEAF, "annotation", "trifid_with_Alt_introns.tsv"))

table(new_trifid$sig)       
table(new_trifid$in_trifid)
table(new_trifid$annotation)
length(unique(new_trifid$intron_coords))
nrow(new_trifid)
```

58 clusters (corresponding to 116 two top introns) were not found in trifid.  

```{r}
table(new_trifid[c("annotation", "sig")])
table(new_trifid[c("sig", "in_trifid")])
new_trifid[new_trifid$gene == 'PEMT',]
```

```{r}
ggplot(new_trifid, aes(x=sig, fill=in_trifid)) + geom_bar(position="dodge")
```
```{r}
ggplot(new_trifid, aes(x=sig, y=mean_norm_score, colour=sig)) + geom_boxplot() + geom_jitter()
```

## Merge sig and not sig

```{r}
old_trifid = rename(old_trifid, gene = gene_name.x)
old_trifid$in_trifid = "in_trifid"
old_trifid$annotation = "gencode"
colnames(old_trifid)[colnames(old_trifid) %in% colnames(new_trifid)]
colnames(old_trifid)[!colnames(old_trifid) %in% colnames(new_trifid)]
to_add = old_trifid[!old_trifid$intron_coords %in% new_trifid$intron_coords,]
nrow(to_add)

trifid = bind_rows(new_trifid, to_add)
tail(trifid)
```

```{r}
ggplot(trifid, aes(x=sig, y=mean_norm_score, colour=sig)) + geom_boxplot() + geom_jitter()
```

Let's simplify these categories...

sigbeige, sig white, pvalue, neither

```{r}
trifid$category = trifid$sig
trifid$category[trifid$category == "white"] = "p-value"
trifid$category[trifid$category == "beige"] = "p-value"
ggplot(filter(trifid, in_trifid == "in_trifid"), aes(x=category, y=mean_norm_score, colour=category)) + geom_violin() + 
  geom_boxplot(width=0.2) #+ geom_jitter()
```
n=257 beige and n=280 white introns are in trifid to be compared like this 
```{r}
ggviolin(filter(trifid, in_trifid == "in_trifid"), x="category", y="mean_norm_score", col="category", trim=T) +
    stat_compare_means(comparisons = list( c("neither","p-value"),
                                        c("sig_white","sig_beige"),
                                          c("sig_beige","neither"),
                                          c("sig_white","neither")) ) +
        stat_compare_means(method = "kruskal.test", label.y=1.5) + 
    geom_boxplot(fill=NA, width=0.2) 
```

I wonder if the number of introns found for the ovalue and neither category could be subsampled to produce similar numbers and verify the statistics?  
The white and beige categories each have ~300 introns in trifid, so lets check.  

```{r}
filt = bind_rows(filter(trifid,category %in% c("sig_white","sig_beige") & in_trifid=="in_trifid"),
                 sample_n(filter(trifid,category == "neither" & in_trifid=="in_trifid"), 280),
                 sample_n(filter(trifid,category == "p-value" & in_trifid=="in_trifid"), 280))
nrow(filt)
```

```{r}
ggviolin(filt, x="category", y="mean_norm_score", col="category", trim=T) +
    stat_compare_means(comparisons = list( c("neither","p-value"),
                                        c("sig_white","sig_beige"),
                                          c("sig_beige","neither"),
                                          c("sig_white","neither")) ) +
        stat_compare_means(method = "kruskal.test", label.y=1.5) + 
    geom_boxplot(fill=NA, width=0.2)  + geom_jitter(aes(colour=category), alpha=0.5) + 
  theme_classic(base_size=18) 
ggsave(file=file.path(figs, "trifid_stats.pdf"), width=7, height=4.5)
```

Those p-values seem more reasonable :)

# Proportion of trifid not founds

we're going to have to beef up these figures... because I haven't kept all of the unfound introns :)

```{r}
ggplot(trifid, aes(x= category,fill =in_trifid == "in_trifid",)) + geom_bar(position="fill")
```


```{r}
gencode_introns = read.delim(file.path(LEAF, "annotation/all_intron_numbers&transcripts.tsv"))
head(gencode_introns)
nrow(filter(gencode_introns, p.adjust < 0.05 & abs(deltapsi) < 0.1)) #36326 introns in low pvalue clusters, but without the deltapsi
nrow(filter(gencode_introns, p.adjust > 0.05)) #97 913 non-significant gencode introns 
```


```{r}
trifid$simple_trifid = trifid$in_trifid
trifid$simple_trifid [trifid$simple_trifid == "cluster_not_in_trifid"] = "not_in_trifid"
prop = group_by(trifid, category, simple_trifid) %>% count()
prop

## we can update the numbers 
prop = data.frame(in_trifid = c(58240,20000,280,257),
                  not_in_trifid = c(39673, 16326, 90, 66),
                  category = c("neither","p-value","sig_white","sig_beige"))
head(prop)
prop = pivot_longer(prop, 1:2)
prop$name = factor(prop$name, levels=c("not_in_trifid", "in_trifid"))
ggplot(prop, aes(x=category, fill= name, y=value)) + geom_bar(stat="identity")
ggplot(prop, aes(x=category, fill= name, y=value)) + geom_bar(stat="identity", position="fill") + labs(y="Proportion")
```

non sig = 58240/ 97913 in trifid/not =  59.4%
in sig cluster = 20000/36326 = 55.1%
sig white = 280/(280+66) = 81.0%
sig_beige = 257/(257+90) = 74.1%


