---
title: "TRIFID_using_extra_introns"
output: html_document
---

Also found out theres a refseq TRIFID table, so for introns with only a refseq id I use the TRIFID score for refseq transcripts.  
still wokring on adding in the alternative transcripts

```{r}
library(biomaRt)
library(tidyr)
library(dplyr)
library(ggplot2)
library(ggrepel)

library(EnhancedVolcano)
```


```{r}
LEAF= "/home/sarahhp/home/rnaseq/six_donor_trans/output/31_leafcutter"
lc = read.delim(file.path(LEAF, "annotation/three_database_info_sig_junctions.tsv"))
lc$strand = gsub(".*_","", lc$cluster_id)
lc = unite(lc, "intron_coords", chr, intron_coords, strand, sep = ":")
lc[lc$gene == "PEMT",]
head(lc)
dim(lc) # some duplications based on gene names/ antisense transcripts.  
```

```{r}
trifid = read.delim("/projects/imb-pkbphil/sp/rnaseq/other_data/trifid/gencode37_trifid_predictions.tsv")
head(trifid)
length(unique(trifid$transcript_id))#104 688

trefseq = read.delim("/projects/imb-pkbphil/sp/rnaseq/other_data/trifid/refseq110_trifid_predictions.tsv")
nrow(trefseq)
head(trefseq)
trefseq$gene_id = as.character(trefseq$gene_id)
trifid = bind_rows(trifid, trefseq)
nrow(trifid)
```

The reason trifid has so few transcripts is only those with translated sequences are included.  

## Convert lc introns to transcript table

```{r}
transcripts = separate_longer_delim(lc, transcript_ids, delim=",")
transcripts$transcript_ids = gsub("\\.[0-9]*","",gsub("rna-","",transcripts$transcript_ids))
head(transcripts) 
dim(transcripts)

transcripts[transcripts$gene == "PEMT",]
```
 some numbers on this
```{r}
sig_trans = filter(transcripts, p.adjust < 0.05 & abs(deltapsi) >= 0.1)
sprintf("Leafcutter significant introns (%d) correspond to %d unique transcripts (represented in %d rows).",
        length(unique(sig_trans$intron_coords)),
        length(unique(sig_trans$transcript_ids)),
        length(sig_trans$transcript_ids))

sprintf("%d of these transcripts are represented in the trifid database.\n",
        sum(unique(sig_trans$transcript_ids) %in% trifid$transcript_id))

transcripts$transcript_ids[transcripts$gene == "PEMT"] %in% trifid$transcript_id
```


```{r}
cat("Unrepresented transcripts include those from genes like: \n")
cat(head(unique(sig_trans$gene[!sig_trans$transcript_ids %in% trifid$transcript_id])))

missing = transcripts[!transcripts$transcript_ids %in% trifid$transcript_id,] #annotate later
table(missing$annotation)
missing[missing$gene == "PEMT",]
```

 Of course refseq and fantom transcripts cannot be annotated by this trifid (which is based on the ensembl annotation); but also many transcripts from gencode cannot be found there either - perhaps they're new transcripts, not protein coding versions.  

## Merge the extra introns
Alt intron = the next highest intron from a cluster with only 1 significant intron.  Aka this intron has deltapsi < 0.1, but may represent the reciprocal to a significant intron that is significant. 

```{r}
alt_introns = read.delim(file.path(LEAF, "annotation/alt_introns_195.tsv"))
alt_trans = separate_longer_delim(alt_introns, transcript_ids, delim=",")
alt_trans$transcript_ids = gsub("\\.[0-9]*","",gsub("rna-","",alt_trans$transcript_ids))
head(alt_trans)

alt_trans$sig = NULL
summary(alt_trans$sig)
```

```{r}
all_trans = rbind(transcripts[colnames(transcripts) %in% colnames(alt_trans)], alt_trans)
nrow(all_trans)
head(all_trans)
#update missing to include non-significant alt introns
missing = all_trans[!all_trans$transcript_ids %in% trifid$transcript_id,]
nrow(missing) #we'll add these in later if necessary
head(missing)
```


## Merge trifid and leafcutter

```{r}
mtrans = merge(all_trans, trifid, by.x ="transcript_ids",by.y="transcript_id")
mtrans = mutate(mtrans, condition = if_else(deltapsi > 0, "beige", if_else(deltapsi < 0, "white", "none") ),
                sig = p.adjust < 0.05 & abs(deltapsi) >= 0.1)

length(unique(mtrans$transcript_ids)) #1655 transcripts in total
length(unique(filter(mtrans, sig) %>% pull(transcript_ids))) #1301 + 82 from significant introns > 0.1 dpsi
```


```{r}
mtrans = arrange(mtrans, desc(trifid_score), desc(norm_trifid_score), p.adjust, desc(abs(deltapsi))) 
head(mtrans[c("gene_name", "trifid_score","norm_trifid_score","deltapsi","p.adjust")], n=20)

mtrans[mtrans$gene == "PEMT",]
```

## summarise transcripts with introns in multiple directions
```{r}
table(mtrans$num_introns)
table(mtrans$condition)
table(paste(mtrans$sig, mtrans$condition))
```

## Histogram

```{r}
ggplot(mtrans) + geom_histogram(aes(x=trifid_score, fill=condition), bins=30)
ggplot(mtrans) + geom_histogram(aes(x=norm_trifid_score, fill=condition), bins=30)
ggplot(filter(mtrans, sig)) + geom_histogram(aes(x=trifid_score, fill=condition), bins=30)
```
## Flags

```{r}
table(mtrans$flags)
table(filter(mtrans, sig) %>% pull(flags))
```
## Correlation between dPSI and TRIFID score?

There is actually, if you look within the same cluster if we have a intron with a higher PSI it tends to have higher functionality - or maybe it just has more transcripts?  
```{r}
ggplot(mtrans, aes(x=deltapsi, y=norm_trifid_score, colour=condition)) + geom_point(size=0.5) +
    geom_text_repel(data= filter(mtrans, abs(deltapsi) >0.3), aes(label=gene))
ggplot(mtrans, aes(x=abs(deltapsi), y=norm_trifid_score, colour=condition)) + geom_point(size=0.5) +
    geom_text_repel(data= filter(mtrans, abs(deltapsi) >0.3), aes(label=gene))

ggplot(mtrans, aes(x=paste(condition, sig), y=norm_trifid_score, fill=condition)) + geom_violin() + geom_boxplot(width=0.5, fill=NA) 
ggplot(mtrans, aes(x=paste(condition, sig), y=norm_trifid_score, fill=condition)) + geom_violin() +
    geom_jitter(size=0.75)
ggplot(mtrans, aes(x=paste(condition, sig), y=norm_trifid_score, colour=condition)) + geom_boxplot() +
    geom_jitter(size=0.75)


#is this true onun-normalised trifid score?
ggplot(mtrans, aes(x=paste(condition, sig), y=trifid_score, fill=condition)) + geom_violin() + geom_boxplot(width=0.5, fill=NA) 
```

Transcripts that can be formed from introns that were enriched in beige, have on average higher TRIFID scores, potentially indicating higher functionality for beige specific isoforms.  This relationship also holds for the un-normalised trifid score as well.  However, on average the white specific transcript are also more functional than the total expressed transcriptome (according to leafcutter).  

## Heatmap transcript vs condition trifid score?  



## Annotate transcript names

```{r}
mart <- useMart(biomart = "ensembl",
  dataset = "hsapiens_gene_ensembl",
   host = "https://sep2019.archive.ensembl.org")

biomaRt::searchAttributes(mart = mart, pattern = "transcript.*name")

annot = getBM(c("external_transcript_name","ensembl_gene_id", "ensembl_transcript_id"), 
              filters = "ensembl_transcript_id",
              values =  mtrans$transcript_ids,
              mart = mart, useCache = F)

head(annot, n=2); dim(annot)

#Add gene names to filt_series
mtrans = merge(mtrans,annot, by.x= "transcript_ids", 
                        by.y = "ensembl_transcript_id", sort=FALSE,
               all.x = T)
#head(mtrans)
remove(annot)
```


## Averaging score across isoforms at each intron
Well they're not going to be different per intron, because the intron is found in the same transcripts in both obv.  We're not comparing two, unless we have two introns from the same gene...


```{r}
summary(mtrans$trifid_score)

head(mtrans)
dim(mtrans)
#mtrans = filter(mtrans, sig)

#unknown gene names get given a ".", change to NA
mtrans = mutate(mtrans, gene = if_else(gene==".", cluster_id, gene))

#this step  averages the score if multiple transcripts are implicated at an intron
introns = group_by(mtrans, intron_coords,condition, deltapsi, p.adjust, gene, cluster_id, annotation) %>% summarise(mean_trifid_score = mean(trifid_score, na.rm = T),
                                                                 mean_norm_score = mean(norm_trifid_score, na.rm = T),
                                                                  median_norm_score = median(norm_trifid_score, na.rm = T),
                                                                 transcripts = paste(transcript_ids, sep=",",collapse=","),
                                                                 transcript_names = paste(external_transcript_name,sep=",", collapse=","),) 
introns = arrange(introns, desc(abs(deltapsi)), desc(mean_trifid_score), p.adjust,)
dim(introns) #78756 introns with annotations
head(introns)
length(unique(introns$intron_coords))

```

## If an intron is not in trifid; report
```{r}
sig_clusters = unique(lc$cluster_id[lc$p.adjust < 0.05 & abs(lc$deltapsi) > 0.1])
summary(sig_clusters %in% introns$cluster_id)#379 clusters have at least 1 trifid annotation; 58 do not
#must have another annotated intron in the cluster, but not have a score already for that intron
to_add = missing[missing$cluster_id %in% sig_clusters & 
                     !missing$intron_coords %in% introns$intron_coords,]
nrow(to_add) #160 of the unannotated trifid transcripts, have an annotated cluster and are for an unscored intron

to_add = mutate(to_add, gene = if_else(gene==".", cluster_id, gene))
to_add = group_by(to_add, intron_coords, deltapsi, p.adjust, gene, cluster_id, annotation) %>% 
    summarise(transcripts = paste(transcript_ids,collapse=","),
              ) 
to_add = mutate(to_add, mean_trifid_score = -0.1,
                        mean_norm_score = -0.1,
                        median_norm_score = -0.1,
                condition=  if_else(deltapsi > 0, "beige", "white"))
head(to_add)
nrow(to_add)
length(unique(to_add$intron_coords))
to_add= distinct(to_add) #managed to duplicate
nrow(to_add)
```

```{r}
all_introns = bind_rows(in_trifid=introns, not_in_trifid=to_add[to_add$cluster_id %in% introns$cluster_id,],
                        cluster_not_in_trifid= to_add[!to_add$cluster_id %in% introns$cluster_id,],
                        .id = "in_trifid" )
nrow(all_introns)
length(unique(all_introns$intron_coords))
```


## Violin plots on *average* TRIFID score

```{r}
all_introns =mutate(all_introns, sig = if_else(p.adjust < 0.05 & deltapsi > 0.1, "sig_beige", 
                                       if_else(p.adjust < 0.05 & deltapsi < -0.1, "sig_white",
                                                       if_else(deltapsi > 0, "beige", 
                                                               if_else(deltapsi < -0, "white", "neither")))) )
table(all_introns$sig)

head(filter(all_introns, startsWith(sig, "sig_")))
head(filter(all_introns, condition == "beige"))


ggplot(all_introns, aes(x=sig, y=mean_norm_score, fill=sig)) + geom_violin() + geom_boxplot(width=0.5, fill=NA) 
ggplot(all_introns, aes(x=sig, y=mean_norm_score, fill=sig)) + geom_violin() +
    geom_jitter(size=0.75)
ggplot(all_introns, aes(x= sig,  y=mean_norm_score,colour=sig)) + geom_boxplot() +
    geom_jitter(size=0.75)
```
The non-averaged score would be appropriate if the number of transcripts per intron was similar between the groups.  Trifid scores (more approprate if we're comparing across gene instead of pairwise between within each gene); has the average score HIGHER in white than beige.  also a nice validation of the psi threshold.  

```{r}
write.table(all_introns, file.path(LEAF, "annotation/trifid_with_Alt_introns.tsv"), sep="\t",quote=F, row.names = F)
```


## Plot the distance

```{r}
head(all_introns)
ggplot(all_introns, aes(x=condition, y=mean_norm_score)) + geom_jitter() + geom_boxplot(fill=NA) + geom_line(aes(group=cluster_id))
```

Can we split this into increased and decreased functionality? Does that make sense.  Should we do a volcano plot instead?  

```{r}
ggplot(all_introns, aes(colour=in_trifid, y=mean_norm_score, x=deltapsi)) + geom_jitter() + 
    geom_line(data=filter(all_introns, abs(deltapsi) > 0.25), aes(group=cluster_id), colour="black", alpha=0.5) + 
    geom_text_repel(data=filter(all_introns, abs(deltapsi) > 0.25), aes(label=gene))
ggplot(all_introns, aes(colour=condition, y=mean_norm_score, x=deltapsi)) + geom_jitter() + 
    geom_line(data=filter(all_introns, abs(deltapsi) > 0.25), aes(group=cluster_id), colour="black", alpha=0.5) + 
    geom_text_repel(data=filter(all_introns, abs(deltapsi) > 0.25), aes(label=gene))
```

Some genes, like PEMT don't have a trifid score from the same cluster. Could be interesting to get the coding status (or appris scores) on here in colours to show why the switch is occuring in some cases - though it would have to be simplified.  Interesting in terms of TRIFID a decent chunk of high deltapsi in beige have low functionality in beige.  Unfortunately HDAC8 and CITED1 are connected because they're in the same cluster.

```{r}
mtrans[mtrans$gene == "PEMT",]
all_trans[all_trans$gene == "PEMT",]
```


## calculate the distance


```{r}
#if num all_introns is 3 drop for this analysis? or can we expand them?
#if clusterid appears more than twice, add 2 to the id?  
#the easiest is for me to average between all_introns. 

paste_uq = function(x){
    return = paste(unique(x), collapse=",")
}
mean_of_3s = filter(all_introns, in_trifid != "cluster_not_in_trifid") %>%
    group_by( cluster_id, condition, p.adjust) %>% summarise(mean_norm_score = mean(mean_norm_score),
                                                                deltapsi = mean(deltapsi),
                                                                gene = paste(unique(gene), collapse=","),
                                                                transcript_names= paste(unique(transcript_names), collapse=","))
head(mean_of_3s)
filter(mean_of_3s, is.na(deltapsi))


trifid_diff = pivot_wider(mean_of_3s, names_from = condition, values_from = c(deltapsi, mean_norm_score, gene), values_fn = list(deltapsi=mean,mean_norm_score=mean, gene=paste_uq ),
                          id_cols=c("cluster_id", "p.adjust"))
head(trifid_diff)

#mean_of_3s[mean_of_3s$cluster_id == "clu_291_-",]
#mean_of_3s[mean_of_3s$cluster_id == "clu_10654_+",]

## If your other intron is not in trifid speak now
## deltapsi we should have from the other table, just the trifid score we could set to -1
filter(trifid_diff, is.na(deltapsi_beige ) | is.na(deltapsi_white )) # just 28 with an intron pair missing thought I captured all of those
```

```{r}
trifid_diff = mutate(trifid_diff, trifid_diff = mean_norm_score_beige-mean_norm_score_white, max_dpsi = max(abs(c(deltapsi_white, deltapsi_beige))))
head(trifid_diff)
```


```{r}
figs = file.path(LEAF, "../../R/leafcutter/plots")
ggplot(trifid_diff, aes(x=max_dpsi, y=trifid_diff, colour=trifid_diff)) + geom_point() + 
    geom_label_repel(data=filter(trifid_diff, max_dpsi > 0.25), aes(label=gene_beige)) + 
    scale_color_gradient2(low="turquoise", mid="black", high="red")


ggplot(trifid_diff, aes(y=max_dpsi, x=trifid_diff, colour=trifid_diff)) + geom_point() + 
    geom_label_repel(data=filter(trifid_diff, max_dpsi > 0.25), aes(label=gene_beige)) + 
    scale_color_gradient2(low="turquoise", mid="black", high="red")
ggsave(file.path(figs, "trifid_difference_v_dpsi.pdf"))

ggplot(trifid_diff, aes(y=-log10(p.adjust), x=trifid_diff, colour=trifid_diff)) + geom_point() + 
    geom_label_repel(data=filter(trifid_diff, p.adjust < 0.01), aes(label=gene_beige)) + 
    scale_color_gradient2(low="turquoise", mid="black", high="red")
ggsave(file.path(figs, "trifid_difference_v_pvalue.pdf"))
```
Hmm.... CITED1 really shouldn't be here because we're comparing across genes which doesn't necessarily make sense.  

```{r}
write.table(trifid_diff, file.path(LEAF, "annotation/trifid_DIFFERENCE_with_Alt_introns.tsv"), sep="\t",quote=F, row.names = F)
```


## Actin changes

```{r}
from_go_term = "ADD3/KANK1/DIXDC1/SVIL/PRKG1/MYO1C/INPP5K/MYO6/DPYSL3/SH3KBP1/ARHGEF10L/ARHGEF2/IQSEC1/FMNL2/BCL6/PALLD/EPB41L3/EMP2/PTPN1/ROCK2/FMN2/STAU2/MICAL3/SPTBN1/FGD4/ATP1A2/CLASP2/RUFY3/AKAP13/PIP5K1C/NCK1/CACNA2D1/BCAR1/WASH3P/CALD1/VPS4A/HIP1R/EHD2"

actin = stringr::str_split_1(from_go_term, "/")
str(actin)

ggplot(filter(trifid_diff, gene_beige %in% actin | gene_white %in% actin), 
       aes(y=-log10(p.adjust), x=trifid_diff, colour=trifid_diff)) + geom_point() + 
    geom_label_repel( aes(label=gene_beige)) + 
    scale_color_gradient2(low="turquoise", mid="black", high="red") + theme_classic(base_size=18)
ggsave(file.path(figs, "actin_trifid_scores.pdf"), width = 10, height = 10)

ggplot(filter(trifid_diff, gene_beige %in% actin | gene_white %in% actin), 
       aes(y=max_dpsi, x=trifid_diff, colour=trifid_diff)) + geom_point() + 
    geom_label_repel( aes(label=gene_beige)) + 
    scale_color_gradient2(low="turquoise", mid="black", high="red") + theme_classic(base_size=18)
ggsave(file.path(figs, "actin_trifid_scores_dpsi.pdf"), width = 10, height = 10)

ggplot(filter(trifid_diff, gene_beige %in% actin | gene_white %in% actin), 
       aes(y=-log10(p.adjust), x=trifid_diff, colour=trifid_diff)) + geom_point() + 
    scale_color_gradient2(low="turquoise", mid="black", high="red")

ggplot(filter(trifid_diff, gene_beige %in% actin | gene_white %in% actin), 
       aes(x=trifid_diff)) + geom_histogram(bins=10)

ggplot(filter(trifid_diff, gene_beige %in% actin | gene_white %in% actin), 
       aes(x=trifid_diff)) + geom_dotplot(bins=10)
```

